src/tmpfiles/tmpfiles.c

```diff
@@ -2006,62 +2006,176 @@ static int glob_item_recursively(Item *i, fdaction_t action) {
+                        // TODO fix this log to give full path like log_unsafe_transition
+                        log_error_errno(r, "rm_rf(%s): %m", name);
                 }
                 return -errno;
         }
+        return -ENOENT;
+}
 
-        return 0;
+static void close_parent(int* parent_fd) {
+        if (*parent_fd != AT_FDCWD) {
+                safe_close(*parent_fd);
+        }
+        *parent_fd = -1;
 }
 
-static int rm_nondir_parents(const char *path) {
-        char *p, *e;
-        int r = -ENOENT;
+/* If child_mode is non-zero, rm_if_wrong_type_safe will be executed for the last path component. */
+static int mkdir_parents_rm_if_wrong_type(mode_t child_mode, const char *path) {
+        _cleanup_(close_parent) int parent_fd = -1;
+        _cleanup_close_ int next_fd = -1;
+        struct stat parent_st;
+        const char *s, *e;
+        int r;
+        size_t path_len;
 
         assert(path);
+        assert((child_mode & ~S_IFMT) == 0);
 
-        // Walk up the path components until one is found with the expected type or there are no more.
-        p = strdupa(path);
-        // If the path doesn't exist, check the next path component.
-        while (r == -ENOENT || r == -ENOTDIR) {
-                e = strrchr(p, '/');
-                if (!e)
-                        return 0;
+        path_len = strlen(path);
+
+        e = strrchr(path, '/');
+        if (!e)
+                return child_mode ? rm_if_wrong_type_safe(child_mode, AT_FDCWD, NULL, path, false) : 0;
+
+        /* Trailing path separators are only allowed if child_mode isn't set. */
+        if ((e - path + 1 == path_len) && child_mode)
+                return -EINVAL;
+
+        /* Get the parent_fd and stat. */
+        parent_fd = AT_FDCWD;
+        if (*path == '/') {
+                parent_fd = open("/", O_NOCTTY | O_CLOEXEC | O_DIRECTORY);
+                // TODO log on error.
+                if (parent_fd < 0)
+                        return -errno;
+                r = fstat(parent_fd, &parent_st);
+        } else {
+                r = stat("", &parent_st);
+        }
+        // TODO log on error.
+        if (r < 0)
+                return -errno;
+
+        /* Check every parent directory in the path, except the last component */
+        e = path;
+        for (;;) {
+                char t[path_len + 1];
+
+                /* Find the start of the next path component. */
+                s = e + strspn(e, "/");
+                /* Find the end of the next path component. */
+                e = s + strcspn(s, "/");
+
+                /* Copy the path component to t so it can be a null terminated string. */
+                memcpy(t, s, e - s);
+                t[e - s] = 0;
+
+                /* Is this the last component? If so, then check the type */
+                if (*e == 0)
+                        return child_mode ? rm_if_wrong_type_safe(child_mode, parent_fd, &parent_st, t, false) : 0;
```

Comment:
It might make sense to return the parent_fd instead of closing it for use in the create operations.